### DISCUSSION TITLE
Doubt in `Lesson 43: Solidity's Free Memory Pointer`

### DISCUSSION THREAD

**Discussion by Basha-dude** [Cyfrin/assembly-evm-opcodes-and-formal-verification-course#17] [General]
hey @EngrPips  in the Lesson:43, @PatrickAlphaC  said that In `memory` 0x00 is `first` slot, 0x20 is `second` slot,0x40 is `fourth` slot  i do not understand how?

**Comment by EngrPips** [comment]
Hey @Basha-dude, seem like you understand it now.

**Comment by Basha-dude** [comment]
Yep😂👍, it is happening always like this 😂

**Reply by EngrPips** [reply]
Indeed, learning happens like that.

**Comment by PatrickAlphaC** [comment]
Not sure if this video helps or makes it harder, let me know!

https://www.youtube.com/shorts/6lpn4bGnEmY

**Reply by PatrickAlphaC** [reply]
Wow, I hate this short. Wtf am I even saying. This video is not helpful ignore it.

**Reply by EngrPips** [reply]
lol, I think it is helpful sir.

**Reply by Basha-dude** [reply]
Yes it is helpful @PatrickAlphaC  ☺️

**Reply by Basha-dude** [reply]
Hey @EngrPips  i don't know what I am doing in invariant testing, are you good at.?  And how long does it take you to understand it , and I will you me do it to other audited contracts, waiting for your reply :-)

**Reply by EngrPips** [reply]
Hello @Basha-dude, I can't outrightly say I am good at invariant testing as I haven't written anyone myself except for what we wrote in Patrick's course, but I have a good understanding of what we did in Patrick's course. Honestly, Like any other thing, you will get better at it the more you do it. There is no magic to being good at things except for constant practice. Imagine someone who has written only 10 invariant test suites to someone who has written over 500 invariant test suites. I mean even a stupid person knows intuitively that the person who has written 500 invariant is better and will understand it more.

**Reply by Basha-dude** [reply]
@EngrPips ,Perfect 😃👍! Will you help me in other contractsl ok whatever I am doing invariant testing to it, whatever you got

**Reply by EngrPips** [reply]
My hands are so tied that I can't promise anything, but I surely wish you success in your pursuit, and I will help with whatever I can.

**Reply by Basha-dude** [reply]
👍, Hey @EngrPips , did you know what is in Patrick's new course (uploading in YouTube) and when it is coming, is it a new one or some new concepts or old  course with new topics inserted?

**Reply by EngrPips** [reply]
Can you share the link on YouTube to see what you are talking about? I don't get informed about courses before they are released, so I most likely don't know.

**Reply by Basha-dude** [reply]
He [tweeted](https://x.com/PatrickAlphaC/status/1812206387805434258?t=bYpQGhLiYqS2VT7BR2a_6g&s=19)

**Reply by EngrPips** [reply]
I don't know what that is. We shall see.

**Reply by Basha-dude** [reply]
Okay 👍

**Reply by PatrickAlphaC** [reply]
It's basically all of Updraft blockchain basics -> advanced foundry combined into one

**Reply by Basha-dude** [reply]
Yep! 😊 I just saw it! @PatrickAlphaC sir!, is there any idea or  something that to make another invariant testing tutorial in future

**Reply by PatrickAlphaC** [reply]
In the security course, we have a more advanced one... That I need to fix. We also go over formal verification in the assembly and FV course which gives more understanding of how invariants work. 

Yeah, I should make more invariant content.

**Reply by Basha-dude** [reply]
That's fantastic  @PatrickAlphaC sir 🤩, I have doubt someone said invariant testing is good, but formal verification is best, and invariant testing is very lesser important that FV, is it true ?

**Reply by PatrickAlphaC** [reply]
Ahhhh so here is the breakdown.

1. Invariant
A property of a system that must hold. For example, in Uniswap, the invariant is "x * y = k"

2. Invariant test
This is any type of test that checks for an invariant. Foundry sort of made this confusing by calling a specific type of invariant testing "invariant testing" which isn't quite right. Really, there are many types of "invariant" tests, even a unit test can be an invariant test. 

3. Fuzzing
Fuzzing is where you fire random inputs at your tests to try to break an invariant. 

4. Stateless fuzzing
This is where you fire random inputs at your tests to try to break an invariant, and your tests don't "hold" state. Basically you call 1 single function with different inputs each time. Basically, you call a function with random data, reset the smart contract, call it with random data, reset the contract, repeat. Each one of the times you call a function with random data and then reset is called a "fuzz run". 

5. Stateful Fuzzing
Stateful fuzzing is what foundry calls "invariant" tests, and it's where you fire random inputs at your tests to try to break an invariant. Stateful fuzzing "holds" state, meaning that your smart contract storage values will hold between "fuzz runs". Basically, you'll fire random data at a smart contract, and then it'll keep it's state, and you fire more data at another function, then more at another function... And you don't reset. This is called "stateful fuzzing". 

6. Formal Verification
This is similar to stateful fuzzing, except, instead of firing random data, you convert your functions into math and then prove an invariant holds. 

# What is most important? 

Here is my cheat sheet for what you should focus on:

1. Stateless fuzzing
2. Stateful fuzzing
3. Formal Verification

_Only_ after you have good fuzz tests should you consider doing formal verification. Fuzzing will be faster, find 99.9999999% of bugs, and be a much better use of time. FV should come last when making a test suite.

**Reply by EngrPips** [reply]
Phenomenal breakdown here.

**Reply by Basha-dude** [reply]
Yeah that's awesome

**Reply by Basha-dude** [reply]
hey @EngrPips  buddy when i run this command `pip3 install virtualenv` given this
```solidity
rebelbash@LAPTOP-QQLIBCSB:~$ pip3 install virtualenv
Defaulting to user installation because normal site-packages is not writeable
Collecting virtualenv
  Downloading virtualenv-20.26.3-py3-none-any.whl (5.7 MB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.7/5.7 MB 2.7 MB/s eta 0:00:00
Collecting filelock<4,>=3.12.2
  Downloading filelock-3.15.4-py3-none-any.whl (16 kB)
Collecting platformdirs<5,>=3.9.1
  Downloading platformdirs-4.2.2-py3-none-any.whl (18 kB)
Collecting distlib<1,>=0.3.7
  Downloading distlib-0.3.8-py2.py3-none-any.whl (468 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 468.9/468.9 KB 764.3 kB/s eta 0:00:00
Installing collected packages: distlib, platformdirs, filelock, virtualenv
Successfully installed distlib-0.3.8 filelock-3.15.4 platformdirs-4.2.2 virtualenv-20.26.3
```

after this i run this  ```virtualenv``` given this
```solidity
rebelbash@LAPTOP-QQLIBCSB:~$ virtualenv
usage: virtualenv [--version] [--with-traceback] [-v | -q] [--read-only-app-data] [--app-data APP_DATA] [--reset-app-data] [--upgrade-embed-wheels] [--discovery {builtin}] [-p py] [--try-first-with py_exe]
                  [--creator {builtin,cpython3-posix,venv}] [--seeder {app-data,pip}] [--no-seed] [--activators comma_sep_list] [--clear] [--no-vcs-ignore] [--system-site-packages] [--symlinks | --copies] [--no-download | --download]
                  [--extra-search-dir d [d ...]] [--pip version] [--setuptools version] [--wheel version] [--no-pip] [--no-setuptools] [--no-wheel] [--no-periodic-update] [--symlink-app-data] [--prompt prompt] [-h]
                  dest
virtualenv: error: the following arguments are required: dest
SystemExit: 2
```

but  in the video `Section 2: Introduction to Formal Verification & Symbolic Execution | Math Master::19. Wsl Python Setup`  after running this command `virtual`  , given this `command not found`

**Reply by EngrPips** [reply]
Hey @Basha-dude, if you type `virtual` in the terminal, you will get `command not found`, but if you type `virtualenv`, you are supposed to get this `usage: virtualenv [--version] [--with-traceback] [-v | -q] [--read-only-app-data] [--app-data APP_DATA] [--reset-app-data] [--upgrade-embed-wheels] [--discovery {builtin}] [-p py] [--try-first-with py_exe]
                  [--creator {builtin,cpython3-posix,venv}] [--seeder {app-data,pip}] [--no-seed] [--activators comma_sep_list] [--clear] [--no-vcs-ignore] [--system-site-packages] [--symlinks | --copies] [--no-download | --download]
                  [--extra-search-dir d [d ...]] [--pip version] [--setuptools version] [--wheel version] [--no-pip] [--no-setuptools] [--no-wheel] [--no-periodic-update] [--symlink-app-data] [--prompt prompt] [-h]
                  dest
virtualenv: error: the following arguments are required: dest
SystemExit: 2`

 run `virtualenv` and let's see what will be the output.

**Reply by Basha-dude** [reply]
Yes same

**Reply by EngrPips** [reply]
Run `pip3 install virtualenv` then run `virtual` Let's see if you will still get that error

**Reply by Basha-dude** [reply]
Hey @EngrPips  it runs perfect as shown in the video, thanks

**Comment by Basha-dude** [comment]
Hey @EngrPips  it runs perfect as shown in the video, thanks

**Reply by EngrPips** [reply]
You are welcome, Friend. Keep crushing it!

**Reply by Basha-dude** [reply]
hey @EngrPips  buddy , when i run this command `certoraRun ./test/invariant-break/formal-verification/certora/conf/FVCatches.conf` 
giving this 
```solidity
(certora-env) rebelbash@LAPTOP-QQLIBCSB:~/MEN-ARE-BRAVE/security-coursesP1-patrick/sc-exploits-minimized$ certoraRun ./test/invariant-break/formal-verification/certora/conf/FVCatches.conf

`java` is not installed. Installing Java version 11 or later will enable faster CVL specification syntax checking before uploading to the cloud.
Compiling src/invariant-break/FormalVerificationCatches.sol...
Compiling src/invariant-break/FormalVerificationCatches.sol to expose internal function information...
`java` is not installed. Installing Java version 11 or later will enable faster CVL specification syntax checking before uploading to the cloud.

The environment variable CERTORAKEY does not contain a Certora key.
To get a free Certora key please visit https://www.certora.com/signup
```


and i added the  `Certora key` as the instructor said in `23. Certora Wsl`, should i do what patrick sir said in the video, or both same

**Reply by EngrPips** [reply]
Hey @Basha-dude, I didn't learn Certora because I plan to stick to halmos as it is easier to write and more straight-forward so I am not sure what could be wrong here, but let's wait to hear from @PatrickAlphaC  as soon as they are available.

**Reply by Basha-dude** [reply]
Yeah ok, thanks for answering

**Reply by EngrPips** [reply]
You are welcome, Friend.

**Reply by Basha-dude** [reply]
hey @EngrPips  buddy will you give me the link of the `decentralized stable coin` of which has `audit data`, i did not find it in the development course of it

**Reply by EngrPips** [reply]
Hello @Basha-dude, Here https://codehawks.cyfrin.io/c/2023-07-foundry-defi-stablecoin is the link for the smart contract audit for the `Decentralized Stable Coin`

**Reply by Basha-dude** [reply]
Thanks

**Reply by EngrPips** [reply]
You are welcome.

**Reply by Basha-dude** [reply]
hey @EngrPips ,did he explain the `invariant testing` in this repo in any video?

**Reply by EngrPips** [reply]
Hello @Basha-dude, I believe he did some brief introduction about fuzzing in the `Decentralize Stable Coin` section https://updraft.cyfrin.io/courses/advanced-foundry/develop-defi-protocol/create-fuzz-tests-handler
